plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+	fw_vendor_id  = g_strdup_printf ("%02x%02x",
+			       fw_buf[fw_header_offset + NVM_VID_OFFSET_MAJOR],
+			       fw_buf[fw_header_offset + NVM_VID_OFFSET_MINOR]);
+	fw_product_id = g_strdup_printf ("%02x%02x",
+			       fw_buf[fw_header_offset + NVM_PID_OFFSET_MAJOR],
+			       fw_buf[fw_header_offset + NVM_PID_OFFSET_MINOR]);
+
+	g_debug ("writing Thunderbolt firmware version %s", fw_version);
+	g_debug ("writing Thunderbolt product-id %s"      , fw_product_id);
+	g_debug ("writing Thunderbolt vendor-id %s"       , fw_vendor_id);
+
+	/* compare vendor-id, product-id between firmware blob and NVM */
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, nvm_buf, NVM_READ_LENGTH,
+					     0, error)) {
+		g_prefix_error (error, "NVM READ error:");
+		return FALSE;
+	}
+	nvm_vendor_id  = g_strdup_printf ("%02x%02x", nvm_buf[NVM_VID_OFFSET_MAJOR],
+				                      nvm_buf[NVM_VID_OFFSET_MINOR]);
+	nvm_product_id = g_strdup_printf ("%02x%02x", nvm_buf[NVM_PID_OFFSET_MAJOR],
+				                      nvm_buf[NVM_PID_OFFSET_MINOR]);
+
+	if (((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) &&
+	    (g_strcmp0 (nvm_vendor_id , fw_vendor_id) != 0 ||
+	    g_strcmp0 (nvm_product_id, fw_product_id) != 0)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Thunderbolt firmware vendor_id %s, product_id %s"
+			     "doesn't match NVM vendor_id %s, product_id %s",
+			     fw_vendor_id, fw_product_id,
+			     nvm_vendor_id, nvm_product_id);
+		return FALSE;
+	}
+
+	/* firmware install */
+	fw_buf += fw_header_offset;
+	fw_blob_size -= fw_header_offset;
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, fw_buf, fw_blob_size, 0,
+					     error))
+		return FALSE;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_NEEDS_ACTIVATION);
+	fu_device_set_version (device, fw_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_setup (FuDevice *device, GError **error)
+{
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *nvm_version = NULL;
+	g_autofree gchar *nvm_vendor_id  = NULL;
+	g_autofree gchar *nvm_product_id = NULL;
+	guchar buf[NVM_READ_LENGTH] = {0};
+
+	if (!fu_dell_dock_usb4_hub_nvm_read (device, buf, NVM_READ_LENGTH, 0, error)) {
+		g_prefix_error (error, "NVM READ error:");
+		return FALSE;
+	}
+	nvm_version    = g_strdup_printf ("%02x.%02x", buf[NVM_VER_OFFSET_MAJOR],
+				                       buf[NVM_VER_OFFSET_MINOR]),
+	nvm_vendor_id  = g_strdup_printf ("%02x%02x" , buf[NVM_VID_OFFSET_MAJOR],
+				                       buf[NVM_VID_OFFSET_MINOR]);
+	nvm_product_id = g_strdup_printf ("%02x%02x" , buf[NVM_PID_OFFSET_MAJOR],
+				                       buf[NVM_PID_OFFSET_MINOR]);
+
+	/* only add known supported thunderbolt devices */
+	name = g_strdup_printf ("TBT-%s%s", nvm_vendor_id, nvm_product_id);
+	if (g_strcmp0 (name, DELL_DOCK_USB4_INSTANCE_ID) == 0)
+		fu_device_add_instance_id (device, name);
+
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
```

Comment:
I just realized, you need to explicitly set vendor name and vendor ID here to match what is in TBT plugin or this won't be compatible to LVFS.

Reply:
Or can you confirm, does the quirk take it?  Maybe it's fine.

Reply:
I think yes if `TBT-********` is found matched from the quirk file.

```
USB4 controller in Dell dock
  DeviceId:             8b61488521efa1bcbf2138f0feb8e075f9ed7f95
  CompositeId:          438ac67676c310b6568b2859ff085270c9228fdf
  Guid:                 27f4ed60-ed53-52a7-ae28-ea87d9a23302 <- USB\VID_8087&PID_0B40&REV_1234
  Guid:                 b3f8b518-ebdd-5969-9b44-6b096eff5b9b <- USB\VID_8087&PID_0B40
  Guid:                 c6e7a14f-f13c-55e5-99de-d59461db1d27 <- USB\VID_8087&PID_0B40&HUB_0000
  Guid:                 bd79ce60-525b-5f39-a3f6-c98c495039ff <- TBT-00d4b071
  Summary:              USB4 controller
  Plugin:               dell_dock
  Protocol:             com.intel.thunderbolt
  Flags:                updatable|require-ac|registered|dual-image|usable-during-update|skips-restart
  Vendor:               Dell Inc.
  VendorId:             USB:0x8087
  VendorId:             TBT:0x00D4
  Version:              32.00
  VersionFormat:        pair
  Icon:                 thunderbolt
  InstallDuration:      46
  Created:              2021-06-19
  PhysicalId:           usb:02:00:03
  LogicalId:            usb4
  FirmwareSizeMin:      262144
  FirmwareSizeMax:      524288
  Priority:             1
  CustomFlags:          skips-restart,require-ac,dual-image
  PossiblePlugin:       dell_dock
  InternalFlags:        is-open

```

Reply:
LVFS uses the literal `Vendor` or the `VendorId` or both? If it's the `VendorID`, must match all from given or partial in it?

Reply:
It matches vendor id. I think this is totally fine now.